packages/block-editor/src/components/border-radius-control/index.js

```diff
@@ -3,54 +3,71 @@
  */
 import {
 	BaseControl,
-	RangeControl,
 	__experimentalParseQuantityAndUnitFromRawValue as parseQuantityAndUnitFromRawValue,
 	__experimentalUseCustomUnits as useCustomUnits,
+	__experimentalVStack as VStack,
+	__experimentalHStack as HStack,
 } from '@wordpress/components';
-import { useState } from '@wordpress/element';
+import { useState, useMemo } from '@wordpress/element';
 import { __ } from '@wordpress/i18n';
 
 /**
  * Internal dependencies
  */
-import AllInputControl from './all-input-control';
-import InputControls from './input-controls';
 import LinkedButton from './linked-button';
 import { useSettings } from '../use-settings';
-import {
-	getAllValue,
-	getAllUnit,
-	hasDefinedValues,
-	hasMixedValues,
-} from './utils';
+import { hasDefinedValues, hasMixedValues } from './utils';
+import SingleInputControl from './single-input-control';
 
 const DEFAULT_VALUES = {
 	topLeft: undefined,
 	topRight: undefined,
 	bottomLeft: undefined,
 	bottomRight: undefined,
 };
-const MIN_BORDER_RADIUS_VALUE = 0;
-const MAX_BORDER_RADIUS_VALUES = {
-	px: 100,
-	em: 20,
-	rem: 20,
-};
+const RANGE_CONTROL_MAX_SIZE = 8;
+const EMPTY_ARRAY = [];
+function useBorderRadiusSizes( presets ) {
+	const defaultSizes = presets?.default ?? EMPTY_ARRAY;
+	const customSizes = presets?.custom ?? EMPTY_ARRAY;
+	const themeSizes = presets?.theme ?? EMPTY_ARRAY;
+
+	return useMemo( () => {
+		const sizes = [
+			{ name: __( 'None' ), slug: '0', size: 0 },
+			...customSizes,
+			...themeSizes,
+			...defaultSizes,
+		];
```

Comment:
This struck me as a little odd. Is this back-to-front as it appears to spread user, theme, then default sizes?

Reply:
I've probably gotten lost along the way but it appears there was custom merging for spacing sizes added in https://github.com/WordPress/gutenberg/pull/61842 while adding support for a `defaultSpacingSizes` option. This approach moved merged spacing sizes under an origin key.

It doesn't appear that has been implemented for radius sizes and might be a factor in some of the bugs encountered in testing.

At this stage, it doesn't look like we'll add default border radii to the core theme.json. Without that, we probably don't need a border-radius equivalent to `defaultSpacingSizes` or custom merging of presets under origin keys.

If I'm wrong about whether we'll add default border radii presets to core theme.json, implementing a `defaultRadiusSizes` option and merging might require a theme.json version bump like the spacing changes needed.

Reply:
🤦 Nevermind me. I see now where the presets are keyed by origin.

My questions now are:

- Should we be supporting a `defaultRadiusSizes` option?
- Do radius sizes need merging in the same fashion as spacing sizes?
- Shouldn't the merge order here be default, theme, then custom/user?

Reply:
Currently, when default radius sizes are defined within `lib/theme.json` (or via `get_core_data` filter) and a theme redefines any of those sizes using the same slug both are included in the `sizes` options calculated here. That issue might be exacerbated if we ever allowed user origin definitions (`customSizes`). 

<img width="470" alt="Screenshot 2024-12-05 at 3 45 22 pm" src="https://github.com/user-attachments/assets/0875670e-7ac3-40d8-998f-298872c430ff">

This explains part of what I was [seeing while testing](https://github.com/WordPress/gutenberg/pull/67544#pullrequestreview-2480331853).

Reply:
> Should we be supporting a defaultRadiusSizes option?

I don't know, wanted to start small. I'll defer to designers

> Do radius sizes need merging in the same fashion as spacing sizes?

What does this mean?

> Shouldn't the merge order here be default, theme, then custom/user?

I think I copied from the spacing, I guess the idea is that the user ones appears first in the list.

Reply:
> I don't know, wanted to start small. I'll defer to designers

Fair enough 👍 

> What does this mean?

I meant that if the different origins contain a preset item with the same slug, the incoming value overrides the original as the [spacing presets do here](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php?plain=1#L4110).

> I think I copied from the spacing, I guess the idea is that the user ones appears first in the list.

That makes sense.

Reply:
> I meant that if the different origins contain a preset item with the same slug, the incoming value overrides the original as the [spacing presets do here](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php?plain=1#L4110).

Why is this done in the server and not in the client for spacing sizes? This seems like it will hide presets from the global styles UI (assuming one can edit the presets...)

Reply:
That's context I don't have yet, I'm afraid. My hunch is that it might be due to the calculation of spacing presets from the `spacingScale` setting.

Reply:
Even that computation shouldn't be done entirely on the server.